### PR TITLE
Homework6

### PR DESCRIPTION
**Pull request by josefschneider**

### CODE REVIEW

File: course5/homework6/string.cc [hunk 1]
```diff
@@ -0,0 +1,106 @@
+
+#include "string.h"
+
+namespace String {
+
+String::String(const String& copy):
+	mContents(std::make_unique<char[]>(mSize)),
+	mSize(copy.mSize)
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: I'm surprised that it works. I would expect mSize to be not initialized yet in line 7.
Isn't valgrind complaining here?

File: course5/homework6/string.cc [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+String::String(const char* charString)
+{
+	*this = charString;
+}
+
+String& String::operator=(const char* lhs)
+{
+	// Find the end of the string
+	while (lhs[mSize] != '\0')
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: aka std::strlen

File: course5/homework6/string.cc [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	mContents = std::move(move.mContents);
+	return *this;
+}
+
+String::String(const char* charString)
+{
+	*this = charString;
+}
+
+String& String::operator=(const char* lhs)
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: It should be rhs, not lhs, right?

File: course5/homework6/string.cc [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+String& String::operator+=(const String& lhs)
+{
+	*this = std::move(*this + lhs);
+	return *this;
+}
+
+char& String::operator[](const int idx)
+{
+	if (static_cast<unsigned>(idx) >= mSize)
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: Why not simply use the same integer type for mSize and idx?

File: course5/homework6/string.cc [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+		throw std::out_of_range("string index out of range");
+	return mContents[idx];;
+}
+
+const char& String::operator[](const int idx) const
+{
+	if (static_cast<unsigned>(idx) >= mSize)
+		throw std::out_of_range("string index out of range");
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: Q: Is std::string doing range checking in operator[]?

File: course5/homework6/string.cc [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	return mContents[idx];;
+}
+
+bool String::operator==(const String& lhs) const
+{
+	if (mSize != lhs.mSize)
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: try this:
```assert(string("") == string());```

File: course5/homework6/string.cc [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+		return false;
+
+	for (size_t idx = 0; idx < mSize - 1; idx++)
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: std::strcmp or std::equal

File: course5/homework6/main.cc [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+#include <string.h>
+#include <cassert>
+
+int main()
+{
+	{
+		// First check if const char* constructor and c_str() method
+		// work as these will be used for tests later on
+		String::String s("abc");
+		assert(::strcmp(s.c_str(), "abc") == 0);
```
Comment (maciekgajewski) [inline course5/homework6/main.cc]: ```std::strcmp```

File: course5/homework6/main.cc [hunk 9]
```diff
@@ -0,0 +1,135 @@
+
+#include "string.h"
+
+#include <iostream>
+#include <string.h>
```
Comment (maciekgajewski) [inline course5/homework6/main.cc]: ```#include <cstring>```

**Review by maciekgajewski — COMMENTED** [review]
Mostly good; there are some corner case when I expect tests to fail.
Have you run it past valgrind?

File: course5/homework6/string.cc (hunk 1 above)
Comment (josefschneider) [reply course5/homework6/string.cc]: I forgot to run Valgrind - all fixed now

File: course5/homework6/string.cc (hunk 3 above)
Comment (josefschneider) [reply course5/homework6/string.cc]: Yup, silly mistake!

File: course5/homework6/string.cc (hunk 2 above)
Comment (josefschneider) [reply course5/homework6/string.cc]: I hadn't realised we were allowed to use cstring methods, this makes it indeed a lot easier!

File: course5/homework6/string.cc (hunk 4 above)
Comment (josefschneider) [reply course5/homework6/string.cc]: To me it made sense to have idx as an int as that is the usual type for an index for operator[]. I've changed it to size_t

File: course5/homework6/string.cc (hunk 5 above)
Comment (josefschneider) [reply course5/homework6/string.cc]: No it isn't but I thought it might be a good addition. I suppose that seeing as none of these throw I should really add noexcept to a lot of these methods?

File: course5/homework6/string.cc (hunk 6 above)
Comment (josefschneider) [reply course5/homework6/string.cc]: Fixed
